### PR TITLE
feat: add support for vote coefficients

### PR DESCRIPTION
**Pull request by vacekj**
This adds support for creating and downloading the vote coefficients as a CSV, which is a requirement for https://github.com/gitcoinco/grants-stack/issues/1475

### CODE REVIEW

**Comment by vacekj** [comment]
@boudra @gravityblast one thing that's weird is that when the coefficient is 0, it's serialized az `0E-9` in the csv. any idea why this might be?

File: src/server.ts [hunk 1]
```diff
@@ -86,16 +89,72 @@ app.get("/data/:chainId/rounds/:roundId/applications.csv", async (req, res) => {
...
-      ...answers,
+      ...answers
     ]);
   }
 
   res.setHeader("content-type", "text/csv");
   res.send(csv.getHeaderString() + csv.stringifyRecords(records));
 });
 
+app.get("/data/:chainId/rounds/:roundId/votecoefficients.csv", async (req, res) => {
```
Comment (boudra) [inline src/server.ts:100]: vote and coefficients are two words, we should split them following the convention, like `passport_scores.json`

```suggestion
app.get("/data/:chainId/rounds/:roundId/vote_coefficients.csv", async (req, res) => {
```

File: src/server.ts [hunk 2]
```diff
@@ -86,16 +89,72 @@ app.get("/data/:chainId/rounds/:roundId/applications.csv", async (req, res) => {
...
+      ...passport_score
+    } = passportScores.find(score => score.address.toLowerCase() === vote.voter.toLowerCase()) ?? {};
+    return {
+      ...vote,
+      ...passport_score,
+      ...evidence
+    };
+  });
+
+  const records = voteCoefficients.map((voteCoefficient) => Object.values(voteCoefficient));
```
[the remark this answers is not in the document]
Comment (boudra) [reply src/server.ts:129]: To add to this, keep in mind that the votes file could potentially have lots of votes, we should probably do the join with passport and convert to CSV record in one pass, instead of walking a big array twice.

[the remark this answers is not in the document]
Comment (vacekj) [reply src/server.ts:129]: Good catch, I was thinking this might be an issue. Will fix.

File: src/server.ts [hunk 3]
```diff
@@ -86,16 +89,72 @@ app.get("/data/:chainId/rounds/:roundId/applications.csv", async (req, res) => {
...
+  /*TODO: type this once we have unified types*/
+  const passportScores: any[] = JSON.parse(data);
+
+  const voteCoefficients = votes.map((vote) => {
+    const {
+      address,
+      evidence,
+      error,
+      ...passport_score
+    } = passportScores.find(score => score.address.toLowerCase() === vote.voter.toLowerCase()) ?? {};
```
[the remark this answers is not in the document]
Comment (hmrtn) [reply src/server.ts:121]: you're right the current implementation's time complexity is O(m * n), which would definitely would have a significant performance impact. A more efficient approach would be to preprocess the passport scores into an object with constant-time lookups.

File: src/server.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (hmrtn) [reply src/server.ts]: fixed in latest push

[the remark this answers is not in the document]
Comment (boudra) [reply src/server.ts]: we're still looping through the array twice, once to join with passport and the second time to convert it to a CSV row, we should do this in one pass

File: src/http/index.ts [hunk 4]
```diff
@@ -87,16 +90,91 @@ app.get("/data/:chainId/rounds/:roundId/applications.csv", async (req, res) => {
...
+  };
+
+  const passportScoresMap = processPassportScores(passportScores);
+
+  const voteCoefficients: any[] = votes.map((vote: any) => {
+    const voter = vote.voter.toLowerCase();
+    return { ...vote, ...passportScoresMap[voter] ?? {} };
+  });
+
+  const records = voteCoefficients.map((voteCoefficient) => [
```
Comment (boudra) [inline src/http/index.ts:132]: We're still unnecessarily looping through the array twice, once to join with passport and the second time to convert it to a CSV row, we should do this in one pass

Comment (gravityblast) [reply src/http/index.ts:132]: @boudra @hmrtn @vacekj 

Maybe we can improve it even more if instead of loading `passport_scores.json`, we load `passport_valid_addresses.json` directly, so we don't need to loop and check the coefficients, they are already the smaller subset of valid addresses.

We can modify `passport/index.ts` to lowercase the addresses, so we don't need to do it again here. 

We can modify `passport/index.ts` to store an object instead of an array, so we load it and it's already in the final shape we want.

We can try with different data structures to see which one has the fastest lookup among Object, Set, and Map.
I benchmarked with a set of 9 millions strings of 42 characters (like eth addresses), and the average lookups with those data structures are:

```
avgSet 0.0011645 ms
avgMap 0.00043525 ms
avgObj 0.000993375 ms
```

wdyt?

Comment (hmrtn) [reply src/http/index.ts:132]: @gravityblast The `passport_valid_addresses.json` approach is a good idea, and from the quick test it appears Map is the fastest. Alternatively, we could cache the passportScoresMap to avoid recalculating it on every request - if we expect it to not be updated too often. 

@boudra It will still be linear but this is true - we can further optimize this to a single map!

Comment (boudra) [reply src/http/index.ts]: I like the idea of writing an object directly! Maybe we could even just store the threshold as the value instead of all that extra data now that we know it's what we need.

But for me this is now good enough to merge if we want to do this later, together with the calculator code that also accesses this data. 

My intention here was only to make sure we're not pushing very resource intensive code unnecessarily, not really optimizing.

Comment (gravityblast) [reply src/http/index.ts]: yes let's try my optimization in another PR!

but we need to fix the loop in this PR to avoid looping more than once.

Comment (hmrtn) [reply src/http/index.ts]: this was resolved, as far as I could tell, in the latest commit!

Comment (boudra) [reply src/http/index.ts]: yeah it looks good! @hmrtn

**Review by boudra — APPROVED** [review]

File: src/test/server.test.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
+import { describe, test, expect } from "vitest";
```
Comment (boudra) [inline src/test/server.test.ts:1]: @hmrtn @vacekj is there any reason we need this extra test file? it looks like it's a duplicate of `test/http/app.test.ts` it has the exact same tests rewritten in slightly different ways

Comment (vacekj) [reply src/test/server.test.ts:1]: I don't think so, it might just be a duplicate from resolving merge conflicts. Can delete in a further PR.

Comment (boudra) [reply src/test/server.test.ts:1]: Awesome I am deleting it in this PR since it was breaking my build: https://github.com/gitcoinco/allo-indexer/pull/55
